### DISCUSSION TITLE
Defining Nodes (And Other Things)

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#18] [Ideas]
## Summary
PDML documents can be defined using arbitrary node names, whatever the author chooses. They may want to standardize the names they've chosen and their meaning, as well as documenting them. PML could be considered an example of this.

I describe one approach below, inspired by other languages in use today. I quite like it, but of course would welcome other approaches, perspectives, or syntax that aligns with PDML's goals.

## Node Definitions
Perhaps a `def` node can be used to define a node's semantics and behaviors.

```
[def node_name]
```

Which can then be used later in the document.
```
[def AwesomeNode]

[node-demo
  [AwesomeNode]
]
```

The above example shows how you could define a node locally for the current document, but with many nodes you may want to move definitions to a separate file. This is already possible using `[u:ins_file]`.

In addition, one should be able to use node definitions defined in a remote location. For example, some organization may want to use PDML for all their markup needs. They have several different groups with their own nomenclature for things related to their business. Each group would define their own node definitions on their group's website and anyone working in that group would pull in the definitions they need.

## Modules
Naturally, one might choose a node name that was used already somewhere else in the world, and they should have the freedom to do so. We'd need to allow creating namespaces to mitigate naming conflicts.

First, let's consider some examples of things that could need namespaces.

- Node definitions
- Custom PDML Extensions
- Schemas
- `u:set` assignments
- `t:type` assignments and maybe other custom types
- Code in `s:def` nodes
- Maybe more..

Therefore, I don't think it's enough to just define the namespace and node name in one go:
```
[def name_space:node_name]
```

Ideally, you could avoid rewriting the namespace by grouping its items.

### Defining

What we call it in this discussion is not as important as the concept I'd like to explain. I'll use a node called `mod` for now, short for "module".

```
[mod namespace_name
  [def AwesomeNode]
  [def another-awesome-node]
]
```

Items defined inside a `mod` node are name-spaced to it. Therefore, a `mod` node must have a name for them to use. Items defined outside of a `mod` node are defined in the global scope. Hence,
```
[def CoolNode]
[mod my-nodes
  [def AwesomeNode]
]


[CoolNode]
[my-nodes:AwesomeNode]
```

Note: You cannot add definitions to a module you didn't define yourself.

Just like `def` nodes, `mod` nodes can be defined in the same file, another file, or a remote file.

Another example of local definition:
```
[mod html
  [def div]
  [def span]
  [def p]
]


[html:div
  [html:p Some text.]
]
[html:span]
```

This would be the default way to use name-spaced items.

### Nesting
In many cases it makes sense to organize items into subgroups as well. Not that they need to be, but it's nice to be able to. So, `mod` nodes can also contain other `mod` nodes.
```
[mod html
  [def div]
  [def span]
  [def p]
  [mod extended-html
    [def tabs]
    [def tab]
  ]
]


[html:div
  [html:extended-html:tabs
    [html:extended-html:tab Tab 1]
    [html:extended-html:tab Tab 2]
    [html:extended-html:tab Tab 3]
  ]
]
```

### Scoping
To make it less verbose, there could also be a `use` node to bring the namespace into scope and remove the need to write the namespace before each of its items.

```
[use html]

[div
  [p Some text.]
]
[span]
```

It should also be possible to only bring part of a module into scope, allowing more control over name conflicts.
```
[mod html
  [def div]
  [def span]
  [def p]
  [mod extended-html
    [def tabs]
    [def tab]
  ]
]
[mod subset-html
  [mod extended-html
    [def tabs]
    [def tab]
  ]
]


[use html:div]
[use html:p]
[use subset-html]

[div
  [extended-html:tabs
    [extended-html:tab]
    [extended-html:tab]
  ]
  
  [p Some text.]
]
```
In the above example, the `extended-html` node comes from `subset-html`, since only the `div` and `p` nodes from `html` are brought into scope while the entire `subset-html` is brought into scope.

#### Concise Scoping
Some module systems have a syntax that let's you avoid repeating yourself. I think PDML's system should offer this as well since it makes things so much nicer, in my opinion.

Consider the following module tree.
```
[mod a
  [def b]
  [mod c
    [def d]
    [def e]
    [def f]
  ]
  [def g]
]
```

You could bring everything (individually) into scope like:
```
[use a:b]
[use a:c:d]
[use a:c:e]
[use a:c:f]
[use g]
```

Or you could also achieve the same thing with:
```
[use
  a:(
    b
    c:(d e f)
  )
  g
]
```
As you can see, `:()` drills down to a nested scope. I chose `( )` as the delimiters here to avoid them being treated as nodes by a parser, i.e. if we used `c:[d e f]`, then `d` might be treated as a node with content of `e f`. And, thanks to the `( )` attribute syntax, it already gets highlighting (for editors supporting it).

#### Domain-Specific Languages
`mod` and `def` nodes together can be used to define a domain-specific language (DSL) or dialect, such as PML or HTML, as well as groups of items or even single items that can be shared and parsed by a PDML parser. Of course, a parser built specifically for a certain dialect would be free to assume their dialect's namespace is automatically in scope.

For example, PML could be defined like:
```
[mod pml
  [def doc]
  ...
]
```

and a PML parser would be free to allow just `[doc]` instead of `[pml:doc]`.

The goal of everything explained up to this point is to allow anyone to define nodes, their semantics and behaviors, share and reuse them, and for a PDML parser to support them without needing a special parser for each new DSL. But, building such a special-purpose parser for a DSL is still an option for those that have the need or desire.

#### Nested `use`
If it's decided that the nods described here (`def`, `mod`, etc.) should also be name-spaced to a built-in PDML namespace, it could start feeling repetitive when defining a lot of items. If it becomes an issue, maybe the `use` node could nest any valid PDML inside it. Any PDML inside it has the specified namespace in scope.

For example, instead of rewriting a `m` namespace everywhere:
```
[m:mod pml
  [m:def doc]
  ...
]

[m:use pml]
...
```

You could instead do:
```
[use m
  [mod pml
    [def doc]
    ...
  ]
]

[m:use pml]
...
```
Notice the `mod` and `def` nodes don't need to repeat the `m` namespace, but the `[m:use pml]` does because it's outside of it.

This might become boilerplate, but maybe still better than the alternative.

## Attributes
I wanted to introduce the concepts above before getting into details on specific things node definitions would define. One of those things need to be attributes. Maybe there could be an `attributes` node that describes what attributes a node can have; maybe their types, defaults, etc.
```
[def link
  [attributes
    ...
  ]
]
```

Some attributes can appear on many nodes, perhaps they're even global. It might be useful to allow a `mod` node to specify attributes that all of its inner node definitions accept.
```
[mod html
  [def div]
  [def span]
  [attributes
    ...
  ]
]
```
I'm not sure how or if this should also be expressible in the global space, i.e. outside of a `mod` or `def` node.

## Other Things
What node definitions would contain might depend on the context in which they're used. i.e. for user interface or for tooling configurations & scripts, but some things are:

- The child nodes it allows
- Validations
- Lifecycle events, and other scripting (think UI components, or even just defining a "button")
- Documentation, which might be useful if text editors could pick up on it
- Schemas
- A template (encapsulated markup as done in web components)
- Styles

## Privacy
People should be able to allow their node definitions (and therefore, modules) to be used by the public. However, is there a need to somehow tweak how much of it is public? An example of what I mean is, we use HTML elements, but we may not have access to how they're defined or if we can override their elements, etc.

## Viability
I haven't implemented a prototype of the concepts described, but I am curious what the performance would look like. I expect it to be a little slower than if you hard code the node names in the parser, but trading that extra speed for the ability for everyone to define their language without writing a parser for it might be worth it.

**Comment by pml-lang** [comment]
Thanks. These are interesting ideas to ponder about.
I will re-read your message and follow up in a few days.

**Comment by pml-lang** [comment]
> The goal of everything explained up to this point is to allow anyone to define nodes, their semantics and behaviors, share and reuse them, and for a PDML parser to support them without needing a special parser for each new DSL.

Yes, this is an essential feature to be implemented in PDML

The current PDML implementation supports _node specifications_ to define the valid nodes and attributes of a PDML document. This approach is used in PML. The obvious drawback of this approach is that _node specifications_ must be defined in source code, which makes it impractical for many users.

Therefore a future PDML version must support schemas (as stated already in another thread). This will allow users to define PDML document validation rules in a sharable text file using PDML syntax. No need to write source code or configure the PDML parser. The schema can be defined together with the data in a single file, or it can be retrieved from another file or URL.
The PDML parser will then read the schema and convert it into _node specifications_, which will then be used to validate PDML documents.

What you are suggesting (if I understand correctly) would be a third way of validation, namely using `def` nodes _within_ a PDML document to describe its nodes. This could surely be useful in some contexts, but I wonder if we need this approach in addition to schemas.

> We'd need to allow creating namespaces to mitigate naming conflicts.

Yes, that's another feature needed in PDML.

Namespaces are already partially implemented, because I needed them for the extension nodes (`[u: ], [t: ], [s: ]`). But namespaces are not documented yet because there are still some decisions to take and improve the implementation. I like your idea of modules too.

> it makes sense to organize items into subgroups as well. Not that they need to be, but it's nice to be able to.

True, but I'm not sure if we should go that far. Maybe later, if users ask for it.

> there could also be a use node to bring the namespace into scope and remove the need to write the namespace before each of its items.
> a syntax that let's you avoid repeating yourself. I think PDML's system should offer this as well since it makes things so much nicer, in my opinion.

Yes, because all kinds of manual repetition must be avoided, in order to stay DRY. A `use` node, as you suggested, would be useful.

> Maybe there could be an attributes node that describes what attributes a node can have; maybe their types, defaults, etc.

This should be covered by schemas too.

Some useful links related to schemas (for inspiration):

- XML
    [Overview](https://en.wikipedia.org/wiki/XML_schema)
    [XML Schema (W3C)](https://en.wikipedia.org/wiki/XML_Schema_(W3C))
    [RELAX NG](https://en.wikipedia.org/wiki/RELAX_NG)
    [Schematron](https://en.wikipedia.org/wiki/Schematron)

- JSON
    [JSON Schema](https://json-schema.org/)

- Other
    [Apache Avro Schema](https://avro.apache.org/docs/1.11.1/specification/)

**Reply by slanden** [reply]
> Therefore a future PDML version must support schemas (as stated already in another thread). This will allow users to define PDML document validation rules in a sharable text file using PDML syntax.

My experience with schemas is in validating data only, not working with XML and similar scenarios. But, reading up on the RELAX NG schema language I saw some similarities to what I'm describing. I think what I've presented means to serve the same purpose.

> The PDML parser will then read the schema and convert it into node specifications, which will then be used to validate PDML documents.

I've avoided use of the term "schema" for two other reasons. First, to avoid drawing connections to schemas in the traditional sense (or, the way I've used them; not counting databases). Second, to avoid thinking about implementation details. Instead, thinking about the problem and what a readable and practical set of features could look like.

> What you are suggesting (if I understand correctly) would be a third way of validation

What I described means to be the single way to define nodes. Whether they're defining an entire DSL, such as HTML, or just defining a node for their own project, such as some fancy "List" component. A generic PDML parser would learn what content a certain node allows after parsing its definition. Then, when the node shows up later in the document the parser "knows" about it just like it would if it was hard-coded.

A specialized parser can always hard-code what the definitions specify. If I want to build a specialized PML parser, and there's a "PML" module, I could take that same file and use it as concise documentation to reference when building the parser.

> I like your idea of modules too.

After reading more on XML namespaces, I should clarify my usage of the term "namespace".

In XML, I think namespaces are something first-class you can add to elements in the actual markup (not just when defining schemas). Then there's `xmlns="http://relaxng.org/ns/structure/1.0"` (which is different from `ns`), and I don't talk about any such equivalent in my proposal. In fact, I read that this form of namespaces is a major point of confusion and some people are against it.

In my proposal, a "namespace" is only created by defining a "module" with the `mod` node. The first child node should be a single word and will become the "namespace" of any nodes defined inside it. Its reason for being is to avoid name conflicts with other people's nodes, not to establish a hierarchy or structure.

Regarding nested modules:
> True, but I'm not sure if we should go that far. Maybe later, if users ask for it.

Yes, I like the idea of each module only being a single level deep.

Back to node definitions and schemas.. In my proposal I didn't mention specific keywords relating to "schemas". The idea though, is that everything used to define a node, including its attributes (and types), child nodes, etc. belong in a `def` node. It could have been `schema`, but again, I avoided the term on purpose.

I'll give an example below. `mod` and `def` node usage is the same as proposed earlier. Any nodes inside of them only convey the point and are not meant as actual proposed items or syntax. The reason is, I haven't thought about what *all* is needed for generic validation, I've only presented a foundation to build on.

Defining HTML's `<meta>` element could look like:
```pdml
[def meta
  [attributes
    charset (pattern = "utf-8/i")
    content (type = string)
    http-equiv (
      [enum
        content-security-policy
        content-type
        default-style
        x-ua-compatible
        refresh
      ]
    )
    name (type = string)
  ]
]

[- Usage -]
[meta (charset = UTF-8)]
[meta (
  name=viewport
  content="width=device-width, initial-scale=1.0"
)]
```

The `meta` node's attributes are `charset, content, http-equiv, name`. It didn't specify anything about content, so `meta` nodes have no content.

**Reply by pml-lang** [reply]
Thanks for this additional info.

To ensure a smooth implementation of PDML schemas and namespaces, I intend to first suggest a specification/implementation in two separate discussion threads.

With regards to PDML document validation, I suggest that we begin with schemas, taking into consideration your insightful suggestions. If required, we can add other methods to validate nodes and attributes at a later stage.

**Reply by slanden** [reply]
I look forward to hearing your thoughts, and it would be helpful in evolving this discussion.

I don't know how many people check these discussions, but it seems like the PML repositories are more active; maybe posting a link back to these discussions from there would bring in some other voices and perspectives.
